reanalysis/hail_filter_and_classify.py

```diff
@@ -0,0 +1,791 @@
+    :return:
+    """
+    return '-'.join(
+        [
+            locus_details.locus.contig.replace('chr', ''),
+            str(locus_details.locus.position),
+            *locus_details.alleles,
+        ]
+    )
+
+
+def extract_comp_het_details(matrix: hl.MatrixTable):
+    """
+    takes the matrix table, and finds compound-hets per sample
+    based on the gene name only
+
+    :param matrix:
+    """
+
+    # set a new group of values as the key, so that we can collect on them easily
+    ch_matrix = matrix.key_rows_by(matrix.locus, matrix.alleles, matrix.class_4_only)
+    ch_matrix = ch_matrix.annotate_cols(
+        hets=hl.agg.group_by(
+            ch_matrix.info.gene_id,
+            hl.agg.filter(ch_matrix.GT.is_het(), hl.agg.collect(ch_matrix.row_key)),
+        )
+    )
+
+    # extract those possible compound het pairs out as a non-Hail structure
+    compound_hets = {}
+
+    # iterate over the hail table rows
+    # find all variant pair permutations which aren't both class 4
+    for row in ch_matrix.select_cols('hets').col.collect():
+
+        # prepare a summary dict for this sample
+        sample_dict = {}
+
+        # iterate over all the `gene: [var1, var2]` structures
+        for gene, variants in dict(row.hets).items():
+
+            # assess each possible variant pairing
+            for var1, var2 in permutations(variants, 2):
+
+                # skip if both are class 4 only - not valuable pairing
+                if var1.class_4_only == 1 and var2.class_4_only == 1:
+                    continue
+
+                # pair the string transformation
+                sample_dict.setdefault(gene, []).append(
+                    [transform_variant_string(var1), transform_variant_string(var2)]
+                )
+
+        # if we found comp hets, add the content for this sample
+        if len(sample_dict) > 0:
+            compound_hets[row.s] = sample_dict
+
+    return compound_hets
+
+
+def filter_matrix_by_ac(
+    matrix_data: hl.MatrixTable, config: Dict[str, Any]
+) -> hl.MatrixTable:
+    """
+
+    :param matrix_data:
+    :param config:
+    :return: reduced MatrixTable
+    """
+
+    # count the samples in the VCF, and use to decide whether to implement
+    # 'common within this joint call' as a filter
+    # if we reach the sample threshold, filter on AC
+    if matrix_data.count_cols() >= config['min_samples_to_ac_filter']:
+        matrix_data = matrix_data.filter_rows(
+            matrix_data.info.AC <= matrix_data.info.AN // config['ac_filter_percentage']
+        )
+    return matrix_data
+
+
+def filter_matrix_by_variant_attributes(matrix_data: hl.MatrixTable) -> hl.MatrixTable:
+    """
+    filter MT to rows with normalised, high quality variants
+    :param matrix_data:
+    :return:
+    """
+
+    # hard filter for quality; assuming data is well normalised in pipeline
+    matrix_data = matrix_data.filter_rows(
+        (
+            matrix_data.filters.length() == 0
```

Comment:
@vladsaveliev  this might be a good time to revisit the filters. In a typical VCF I would expect the 'filters' to contain 'PASS', or a reason for failure. In the post-VQSR data I'm using, `PASS` is implicit (empty set of filters), and I'm currently ignoring the scenario `FILTERS=VQSR` and `AS_FilterStatus=PASS`.

I'm going to open this up to 3 passing conditions:
1. filters = empty
2. filters = PASS
3. filters = VQSR & AS_FilterStatus=PASS

Does this seem like a reasonable idea? I'm not sure where in the pipeline the Filters=PASS disappear, but it could be some point during the joint-call/recalibration steps

Reply:
My thinking here is that with the current setup, a Singleton VCF would likely be 100% filtered out, as I would have expected PASS variants to have the `filters` as `PASS`

Reply:
Quick revert here, from the Hail docs, FILTERS=PASS is represented as an empty set, so filter 2. above is impossible in MT. Logic allows for empty, and VQSR/PASS, I'll see how many variants are retained this way